projects/driver_controls/src/pedal_calibration.c

```diff
@@ -0,0 +1,155 @@
+#include "pedal_calibration.h"
+#include "string.h"
+
+// Initializes the calibration storage and configures ADC to start reading.
+StatusCode pedal_calibration_init(PedalCalibrationStorage *storage, Ads1015Storage *ads1015_storage,
+                                  Ads1015Channel channel_a, Ads1015Channel channel_b) {
+  memset(storage, 0, sizeof(*storage));
+  storage->ads1015_storage = ads1015_storage;
+  storage->channel_a = channel_a;
+  storage->channel_b = channel_b;
+  status_ok_or_return(ads1015_configure_channel(ads1015_storage, channel_a, true, NULL, NULL));
+  status_ok_or_return(ads1015_configure_channel(ads1015_storage, channel_b, true, NULL, NULL));
+  return STATUS_CODE_OK;
+}
+
+// Given the state (full throttle/brake), it finds the min and max raw readings of each channel.
+// The function samples as many times as PEDAL_CALIBRATION_NUM_SAMPLES.
+StatusCode pedal_calibration_get_band(PedalCalibrationStorage *storage,
+                                      PedalCalibrationState state) {
+  int16_t *min_a = &storage->band[PEDAL_CALIBRATION_CHANNEL_A][state].min;
+
+  int16_t *max_a = &storage->band[PEDAL_CALIBRATION_CHANNEL_A][state].max;
+
+  int16_t *min_b = &storage->band[PEDAL_CALIBRATION_CHANNEL_B][state].min;
+
+  int16_t *max_b = &storage->band[PEDAL_CALIBRATION_CHANNEL_B][state].max;
+
+  int16_t reading_a;
+  int16_t reading_b;
+
+  // Read once first to initialize the max and min.
+
+  status_ok_or_return(ads1015_read_raw(storage->ads1015_storage, storage->channel_a, &reading_a));
+  *min_a = reading_a;
+  *max_a = reading_a;
+
+  status_ok_or_return(ads1015_read_raw(storage->ads1015_storage, storage->channel_b, &reading_b));
+  *min_b = reading_b;
+  *max_b = reading_b;
+
+  // Continue reading to determine the max and min occured reading.
+  for (uint16_t i = 0; i < PEDAL_CALIBRATION_NUM_SAMPLES; i++) {
+    status_ok_or_return(ads1015_read_raw(storage->ads1015_storage, storage->channel_a, &reading_a));
```

Comment:
Isn't there the possibility of sampling too quickly by polling? Since you're always reading buffered data, you might get duplicate results if you sample faster than the ADC provides data.

Also, you seem to be doing the same thing for both channels - this could probably just be done in a loop or a function :)

Reply:
Should I throw in delays? I'm guessing there aren't so many things going on in calibration time that delay would block. Making this periodic with a timer is possible too but I have to make sure it doesn't exit the function, because then the user would receive msgs that should proceed to next step. Delay is much easier. What do you think is better?

Reply:
No, both methods rely on knowledge of the data rate (which may change). Why don't you just use the callbacks?